### PR TITLE
gh-116738: Make _csv module thread-safe

### PR DESCRIPTION
**Pull request by aisk**
This extension module has some module state values which may be read or written in different threads:

- `field_limit`: A simple integer value to store the maximum field size for the parser;
- `dialects`: A dict value to store all the dialects by string names.

## `field_limit`

There is a public function `field_size_limit` that can read and write it, <del>using a critical section to protect it. For other internal usages,</del> `_Py_atomic_load_long` and `_Py_atomic_store_long` are used to make it thread safe.

## `dialects`

There are three public functions `get_dialect`, `unregister_dialect` and `list_dialects` that can access this field. <del>For the first two functions, a critical section is used to protect it.</del>

And the `list_dialects` function just returns the dict's keys with `PyDict_Keys`. I think its result is readonly and `PyDict_Keys` itself [is thread safe](https://github.com/python/cpython/blob/b624490deedc37d02e5f29d19ff5e222df3a64dd/Objects/dictobject.c#L3309-L3311), so we don't need to do anything.
Other internal usages of `dialects` involve accessing its value with `get_dialect_from_registry`, which simply calls `PyDict_GetItemRef` and returns the result. I think we don't need to change it too.

**Update**: I think these APIs is just call `dict` APIs which is already thread-safe, and there are no other state changes inside these APIs, and no intermediate variables need to protect in these functions, so we don't need add locks on them.

* Issue: gh-116738

### CODE REVIEW

File: Modules/_csv.c [hunk 1]
```diff
@@ -1582,7 +1588,10 @@ csv_register_dialect(PyObject *module, PyObject *args, PyObject *kwargs)
     dialect = _call_dialect(module_state, dialect_obj, kwargs);
     if (dialect == NULL)
         return NULL;
-    if (PyDict_SetItem(module_state->dialects, name_obj, dialect) < 0) {
+    Py_BEGIN_CRITICAL_SECTION(module_state->dialects);
+    res = PyDict_SetItem(module_state->dialects, name_obj, dialect);
```
Comment (corona10) [inline Modules/_csv.c]: https://github.com/python/cpython/blob/51aefc5bf907ddffaaf083ded0de773adcdf08c8/Objects/dictobject.c#L2474-L2484
https://github.com/python/cpython/blob/51aefc5bf907ddffaaf083ded0de773adcdf08c8/Objects/dictobject.c#L2458-L2465
`PyDict_SetItem` internally set a critical section no?

File: Modules/_csv.c [hunk 2]
```diff
@@ -702,10 +702,15 @@ parse_grow_buff(ReaderObj *self)
 static int
 parse_add_char(ReaderObj *self, _csvstate *module_state, Py_UCS4 c)
 {
-    if (self->field_len >= module_state->field_limit) {
+#ifdef Py_GIL_DISABLED
+    uint32_t field_limit = _Py_atomic_load_int32(&module_state->field_limit);
```
Comment (corona10) [inline Modules/_csv.c]: Please use
FT_ATOMIC_XXXX macro as possible

https://github.com/python/cpython/blob/51aefc5bf907ddffaaf083ded0de773adcdf08c8/Include/internal/pycore_pyatomic_ft_wrappers.h#L13

If the new addition is needed, please add it.

File: Modules/_csv.c (hunk 1 above)
Comment (aisk) [reply Modules/_csv.c]: Yes, thank you for noticing! I found that all these public functions are just reading or writing the *dialect* dict, via the C APIs which are already thread-safe, and there is no more state changes inside these functions. So I think we don't need any more operations on it.

Comment (corona10) [reply Modules/_csv.c]: So why ``Py_BEGIN_CRITICAL_SECTION(module_state->dialects);`` is needed?

Comment (aisk) [reply Modules/_csv.c]: Oh, sorry, I didn't make it clear. I'm just waiting for your confirmation to see if I'm right about the last comment. Will change it now. ;)

File: Modules/_csv.c [hunk 3]
```diff
@@ -1582,7 +1585,8 @@ csv_register_dialect(PyObject *module, PyObject *args, PyObject *kwargs)
     dialect = _call_dialect(module_state, dialect_obj, kwargs);
     if (dialect == NULL)
         return NULL;
-    if (PyDict_SetItem(module_state->dialects, name_obj, dialect) < 0) {
+    res = PyDict_SetItem(module_state->dialects, name_obj, dialect);
```
Comment (corona10) [inline Modules/_csv.c]: I think that we don't have to use `res` anymore. :)

File: Modules/_csv.c [hunk 4]
```diff
@@ -34,7 +35,7 @@ typedef struct {
     PyTypeObject *dialect_type;
     PyTypeObject *reader_type;
     PyTypeObject *writer_type;
-    long field_limit;   /* max parsed field size */
+    int32_t field_limit;   /* max parsed field size */
```
Comment (corona10) [inline Modules/_csv.c]: Out of curiosity, why is this change needed?

Comment (aisk) [reply Modules/_csv.c]: I saw there is not a `_Py_atomic_load_long` so I changed it to `int32_t`, but now I think we can just add it after you mentioned it, thanks for noticing!

File: Include/cpython/pyatomic_msc.h [hunk 5]
```diff
@@ -355,6 +355,14 @@ _Py_atomic_exchange_int(int *obj, int value)
                                            (int32_t)value);
 }
 
+static inline int
+_Py_atomic_exchange_long(long *obj, long value)
+{
+    _Py_atomic_ASSERT_ARG_TYPE(int32_t);
+    return (int)_Py_atomic_exchange_int32((int32_t *)obj,
+                                           (int32_t)value);
```
Comment (aisk) [inline Include/cpython/pyatomic_msc.h]:
```suggestion
                                          (int32_t)value);
```

File: Modules/_csv.c [hunk 6]
```diff
@@ -14,6 +14,7 @@ module instead.
 #endif
 
 #include "Python.h"
+#include "pycore_pyatomic_ft_wrappers.h"  // FT_ATOMIC_LOAD_LONG
```
Comment (aisk) [inline Modules/_csv.c]:
```suggestion
#include "pycore_pyatomic_ft_wrappers.h"
```

**Review by corona10 — APPROVED** [review]
LGTM, Sorry for the late.

@colesbury Do you have any comments about this?

**Review by colesbury — COMMENTED** [review]
Sorry for the long delay in reviewing this. Two requests:

* Let's change `field_limit` to a `Py_ssize_t`. That seems more appropriate here and avoids the need for adding new atomic bindings. I'd prefer to avoid adding additional atomic bindings because its more code to maintain and its easy to add subtle bugs in the atomic bindings. Additionally, `field_limit` is compared to `field_len`, which is already a `Py_ssize_t`.
* Let's use relaxed atomics for reading and writing `field_limit`, e.g., `FT_ATOMIC_STORE_SSIZE_RELAXED`. Relaxed ordering is fine for these sorts of global settings.

For the data type change:
* `%ld` -> `%zd`
* `PyLong_AsLong` -> `PyLong_AsSsize_t`

**Comment by aisk** [comment]
> Sorry for the long delay in reviewing this. Two requests:
> 
>     * Let's change `field_limit` to a `Py_ssize_t`. That seems more appropriate here and avoids the need for adding new atomic bindings. I'd prefer to avoid adding additional atomic bindings because its more code to maintain and its easy to add subtle bugs in the atomic bindings. Additionally, `field_limit` is compared to `field_len`, which is already a `Py_ssize_t`.
> 
>     * Let's use relaxed atomics for reading and writing `field_limit`, e.g., `FT_ATOMIC_STORE_SSIZE_RELAXED`. Relaxed ordering is fine for these sorts of global settings.
> 
> 
> For the data type change:
> 
>     * `%ld` -> `%zd`
> 
>     * `PyLong_AsLong` -> `PyLong_AsSsize_t`

Thanks for the review, updated!

File: Modules/_csv.c [hunk 7]
```diff
@@ -1652,18 +1654,18 @@ _csv_field_size_limit_impl(PyObject *module, PyObject *new_limit)
...
-        module_state->field_limit = PyLong_AsLong(new_limit);
-        if (module_state->field_limit == -1 && PyErr_Occurred()) {
-            module_state->field_limit = old_limit;
+        Py_ssize_t new_limit_value = PyLong_AsSsize_t(new_limit);
+        if (new_limit_value == -1 && PyErr_Occurred()) {
             return NULL;
         }
+        FT_ATOMIC_STORE_SSIZE_RELAXED(module_state->field_limit, new_limit_value);
     }
     return PyLong_FromLong(old_limit);
```
Comment (Fidget-Spinner) [inline Modules/_csv.c]: I think you should be using https://docs.python.org/3/c-api/long.html#c.PyLong_FromSsize_t here instead.

**Review by kumaraditya303 — APPROVED** [review]
